Fix #20334: Redirect to login page on error in preferences page

GuilhermeRoLopes opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #20334.
2. This PR does the following: 
When the user goes to the preferences page without being logged in, instead of showing the error page, the user is redirected to the login page. After the login is made,  it returns automatically to the preferences page where information about the user account is shown normally. This was achieved by modifying the code in preferences-page-root-component.ts. If an error happens when the method validateCanManageOwnAccount() is executed, router.navigate is called with the PAGES_REGISTERED_WITH_FRONTEND.LOGIN.ROUTE constant from AppConstants and "/preferences" as the return URL. One test case in preferences-page-root-component.spec.ts was also modified since it previously checked if the error page was shown in case of error. Now, it is supposed to check if the user is correctly redirected to the login page in case of error.
3. The original bug occurred because: The preferences page would always handle errors by displaying an error page when there was an issue while executing the validateCanManageOwnAccount() method. 

## Essential Checklist

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.

-->

https://github.com/oppia/oppia/assets/119136653/166f7ae8-4172-498b-9974-f2e9ceb88a58

https://github.com/oppia/oppia/assets/119136653/36a274e9-a728-4f7a-a81c-f95569a0a321

<!--

Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

GuilhermeRoLopes committed:
Redirect to login page on error

The changes were done in
preferences-page-root.component.ts and

preferences-page-root-componente.spec.ts:

Previously, the preferences page would
handle errors by showing an
error page when there was an issue with
validating the user's ability to
manage their own account.

This was done in the case that an error
happened
when executing the
'validateCanManageOwnAccount' method.
Now, instead of
showing an error page, the user is
redirected to the login page. This is
achieved by modifying the code in
preferences-page-root.component.ts and
calling 'this.router.navigate' when an
error happens, instead of
'errorPageIsShown = true'. The login
page route is constructed using the

'AGES_REGISTERED_WITH_FRONTEND.LOGIN.ROUTE'
constant from AppConstants.

Additionally, a 'return_url' query
parameter is passed to the login route
because the user is supposed to
automatically return to the preferences
page after the login is made.

In the

preferences-page-root-componente.spec.ts
one test case was modified:

The test case 'should initialize and
show error page when server respond
with error' was updated and renamed
to 'should initialize and redirect to
login page when server respond
with error'. Previously, this test case
checked that the error page would
be shown in case of an error. Now, it
checks that the user is
redirected to the login page in case of
an error with the correct route and
'return_url' query parameter and also
checks that neither the preferences
page or
the error page was shown, since the
user was redirected to the login
page.

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes commented:
@seanlip PTAL

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

seanlip commented:
Hi @GuilhermeRoLopes, thanks! From the product perspective it looks good, just noted one concern -- around 0:41  in your video, there is a console error. Is it an existing bug? If so, could you point to the issue that tracks it? If not, it probably needs to be looked into and fixed.

seanlip commented:
Also @GuilhermeRoLopes are there other pages in Oppia that would benefit from similar treatment (like learner dashboard etc.) and would it be possible to make a similar fix for them too?

Finally, please leave a comment on the original issue as well so that we can assign you to it, thanks. (And in future please follow the guidance at https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#choosing-a-good-first-issue to provide an explanation of what your PR will do (with names of files you're changing, what you plan to change in each file, etc.), before taking up an issue. Thanks!

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes commented:
> Hi @GuilhermeRoLopes, thanks! From the product perspective it looks good, just noted one concern -- around 0:41 in your video, there is a console error. Is it an existing bug? If so, could you point to the issue that tracks it? If not, it probably needs to be looked into and fixed.

Thank you for your response, @seanlip. I also added a before-video and after testing without my code changes i do think it is an existing bug. I have also found the issue that is related to that bug: #19746.

GuilhermeRoLopes commented:
> Also @GuilhermeRoLopes are there other pages in Oppia that would benefit from similar treatment (like learner dashboard etc.) and would it be possible to make a similar fix for them too?
> 
> Finally, please leave a comment on the original issue as well so that we can assign you to it, thanks. (And in future please follow the guidance at https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#choosing-a-good-first-issue to provide an explanation of what your PR will do (with names of files you're changing, what you plan to change in each file, etc.), before taking up an issue. Thanks!

@seanlip Regarding this question, I found that many files in the codebase display an error page if the validation method called before throws an error. The corresponding spec files typically include a test like 'should initialize and show error page when server responds with error,' which could also be updated accordingly. For instance, delete-account-page-root.component.ts and feedback-updates-page-root.component.ts follow the same logic, showing an error page if the validateCanManageOwnAccount() method fails.

There are also other files with similar logic that call different validation methods such as validateAccessToBlogAuthorProfilePage(), validateAccessToBlogHomePage(), doesLearnerGroupExist(), and doesProfileExist(), all of which display the error page if the method returns an error. Depending on the page, it might make sense to redirect the user to another page in case of error by calling router.navigate() with a return_url parameter or not. Because of this, i do believe a similar fix could be applied where needed, at least in these files that follow the same logic.

seanlip commented:
OK, thanks @GuilhermeRoLopes! Sounds good, I think in general if a page requires log-in to access then we should redirect the user to do that, rather than just 404. The exception is e.g. things like private exploration pages where the access control depends on some parameter in the URL (like the exploration ID) -- in those cases, OK to just show 404 if the resource is not available to the user in their current state.

Would you be up for extending this PR to cover more of those pages, or filing an issue that lists those pages and explains what to do so that someone else can take it up? I think it would be nice to fix this issue generally if possible.

Thanks!

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes commented:
> OK, thanks @GuilhermeRoLopes! Sounds good, I think in general if a page requires log-in to access then we should redirect the user to do that, rather than just 404. The exception is e.g. things like private exploration pages where the access control depends on some parameter in the URL (like the exploration ID) -- in those cases, OK to just show 404 if the resource is not available to the user in their current state.
> 
> Would you be up for extending this PR to cover more of those pages, or filing an issue that lists those pages and explains what to do so that someone else can take it up? I think it would be nice to fix this issue generally if possible.
> 
> Thanks!

Thank you for the response, @seanlip. I would prefer to keep this pull request focused only on the issue with this page and eventually create an issue that references all the other pages where the same type of fix could be applied, along with an explanation of the fix, if thats not a problem.

Thanks!

seanlip commented:
@GuilhermeRoLopes Yup, sounds fine to me, thanks. In that case, please could you go ahead and file the issue while waiting for review from @kevintab95?

Thanks!

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes commented:
> @GuilhermeRoLopes Yup, sounds fine to me, thanks. In that case, please could you go ahead and file the issue while waiting for review from @kevintab95?
> 
> Thanks!

Hi @seanlip, created the issue #20430. I pretend to keep updating the issue comment in case there is more pages where a similar fix could be applied.

Thanks!

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes commented:
Hi @kevintab95,

Could you please review the code changes in this pull request? It would be appreciated.

Thanks!

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

kevintab95 reviewed (approved):
LGTM, thanks @GuilhermeRoLopes! Sorry for the delay.

Commit:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes committed:
Merge branch 'develop' into preferences-error-redirect-to-login

GuilhermeRoLopes commented:
@kevintab95  Thank you very much. Since there are no further issues to be addressed in this PR and the CI tests do pass, could you please merge this pull request when possible or should i mention someone else?

Thanks!
